### PR TITLE
Enhancement: Make analytics reports compatible with Dokan

### PR DESCRIPTION
**Pull request by mrabbani**
### All Submissions:

* [ ] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [ ] My code satisfies feature requirements
* [ ] My code is tested
* [ ] My code passes the PHPCS tests
* [ ] My code has proper inline documentation
* [ ] I've included related pull request(s) (optional)
* [ ] I've included developer documentation (optional)
* [ ] I've added proper labels to this pull request

### Related Pull Request(s)

* Full PR Link

### Closes 

* Closes #

### How to test the changes in this Pull Request:

* Steps or issue link

### Changelog entry

*Title*

Detailed Description of the pull request. What was previous behaviour
and what will be changed in this PR.

### Before Changes

Describe the issue before changes with screenshots(s).

### After Changes

Describe the issue after changes with screenshot(s).

### Feature Video (optional)

Link of detailed video if this PR is for a feature.

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

## Summary by CodeRabbit

- **New Features**
  - Enhanced dependency management with improved autoloading for third-party packages.
  - Introduced advanced analytics reporting tailored for multi-vendor marketplaces, including stock and order statistics.
  - Added new service providers for managing common and analytics-related services, improving modularity.
  - Implemented a new database table for order statistics to enhance data management and reporting.
  - Introduced classes for filtering and customizing WooCommerce analytics queries, enhancing reporting capabilities.
  - Streamlined order processing by removing child order management functionality.
  - Added a comprehensive new documentation section for managing services within a service container, including registration and retrieval processes.

- **Bug Fixes**
  - Disabled specific hooks related to WooCommerce order processing and analytics to prevent unintended behavior.

- **Documentation**
  - Updated documentation to reflect new features and provide usage scenarios for the analytics system.
  - Added detailed instructions for registering and managing services in the service container, including various functionalities.

- **Chores**
  - Improved modularity and maintainability through refactoring of the service provider architecture.

### CODE REVIEW

File: includes/Analytics/Reports/Orders/Stats/DataStore.php [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			array(
+				'order_id'           => $order->get_id(),
+				'seller_id'          => (int) $order->get_meta( '_dokan_vendor_id' ),
+				'order_type'        => (int) ( ( new OrderType() )->get_type( $order ) ),
+				'seller_earning'     => $order->get_meta( '_seller_earning' ),
+				'seller_gateway_fee' => $order->get_meta( '_seller_gateway_fee' ),
+				'seller_discount'    => $order->get_meta( '_seller_discount' ),
+				'admin_commission'   => $order->get_meta( '_admin_commission' ),
+				'admin_gateway_fee'  => $order->get_meta( '_admin_gateway_fee' ),
+				'admin_discount'     => $order->get_meta( '_admin_discount' ),
```
Comment (mrabbani) [inline includes/Analytics/Reports/Orders/Stats/DataStore.php:120]: Need to reconcile for getting these data.

File: includes/Analytics/Reports/Orders/Stats/ScheduleListener.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    /**
+     * Register hooks for WooCommerce analytics order events.
+     *
+     * @return void
+     */
+    public function register_hooks(): void {
+        add_action( 'woocommerce_analytics_update_order_stats', [ $this, 'sync_dokan_order' ] );
+        add_action( 'woocommerce_before_delete_order', [ $this, 'delete_order' ] );
```
Comment (mrabbani) [inline includes/Analytics/Reports/Orders/Stats/ScheduleListener.php:28]: Should we delete stats?

**Yes**
- Report will be mismatch with the `dokan_vendor_balance`.

**No**
- Report will not be synced as per WC stats data.

File: includes/Analytics/Reports/WcSqlQuery.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+class WcSqlQuery extends SqlQuery {
+    /**
+     * Update the default value of $handling from "unfiltered" to an empty string,
+     * allowing filters to be applied to the SQL clauses.
+     *
+     * @param string $type Type of SQL clause (e.g., SELECT, WHERE).
+     * @param string $handling Optional. The handling mode for the clause. Defaults to an empty string.
+     * @return string The SQL clause.
+     */
+    protected function get_sql_clause( $type, $handling = '' ) {
```
Comment (mrabbani) [inline includes/Analytics/Reports/WcSqlQuery.php:16]: Default params should be a valid value

File: includes/Install/Installer.php [hunk 4]
```diff
@@ -536,4 +537,26 @@ private static function parse_update_notice( $content, $new_version ) {
 
         return wp_kses_post( $upgrade_notice );
     }
+
+    protected function create_dokan_order_stats_table() {
```
Comment (mrabbani) [inline includes/Install/Installer.php:541]: What will be happened for exist Dokan User?

File: includes/Analytics/Reports/Orders/Stats/DataStore.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+		if ( ! OrderUtil::is_order( $post_id, array( 'shop_order', 'shop_order_refund' ) ) ) {
+			return;
+		}
+
+		// Retrieve customer details before the order is deleted.
+		$order       = wc_get_order( $order_id );
+		$customer_id = absint( CustomersDataStore::get_existing_customer_id_from_order( $order ) );
+
+		// Delete the order.
+		$wpdb->delete( self::get_db_table_name(), array( 'order_id' => $order_id ) );
```
Comment (mrabbani) [inline includes/Analytics/Reports/Orders/Stats/DataStore.php:173]: What will be happened if table does not exist for existing Dokan?

File: includes/Analytics/Reports/Orders/Stats/DataStore.php [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+				'%f',
+				'%f',
+				'%f',
+				'%f',
+				'%f',
+				'%f',
+			);
+
+			// Update or add the information to the DB.
+			$result = $wpdb->replace( $table_name, $data, $format );
```
Comment (mrabbani) [inline includes/Analytics/Reports/Orders/Stats/DataStore.php:140]: What will be happened if table does not exist for existing Dokan?

File: includes/Order/MiscHooks.php [hunk 7]
```diff
@@ -23,17 +23,17 @@ class MiscHooks {
...
+        // add_filter( 'woocommerce_analytics_orders_select_query', [ $this, 'trim_child_order_for_analytics_order' ] );
+        // add_filter( 'woocommerce_analytics_update_order_stats_data', [ $this, 'trim_child_order_for_analytics_order_stats' ], 10, 2 );
 
         // remove customer info from order export based on setting
-        add_filter( 'dokan_csv_export_headers', [ $this, 'hide_customer_info_from_vendor_order_export' ], 20, 1 );
+        // add_filter( 'dokan_csv_export_headers', [ $this, 'hide_customer_info_from_vendor_order_export' ], 20, 1 );
 
-        add_filter( 'woocommerce_rest_prepare_shop_order_object', [ $this, 'add_vendor_info_in_rest_order' ], 10, 1 );
-        add_filter( 'wp_count_posts', [ $this, 'modify_vendor_order_counts' ], 10, 1 ); // no need to add hpos support for this filter
+        // add_filter( 'woocommerce_rest_prepare_shop_order_object', [ $this, 'add_vendor_info_in_rest_order' ], 10, 1 );
```
Comment (mrabbani) [inline includes/Order/MiscHooks.php:35]: We may have to redo this hoook.

File: includes/Order/MiscHooks.php [hunk 8]
```diff
@@ -23,17 +23,17 @@ class MiscHooks {
...
 
         // Exclude suborders in woocommerce analytics.
-        add_filter( 'woocommerce_analytics_orders_select_query', [ $this, 'trim_child_order_for_analytics_order' ] );
-        add_filter( 'woocommerce_analytics_update_order_stats_data', [ $this, 'trim_child_order_for_analytics_order_stats' ], 10, 2 );
+        // add_filter( 'woocommerce_analytics_orders_select_query', [ $this, 'trim_child_order_for_analytics_order' ] );
+        // add_filter( 'woocommerce_analytics_update_order_stats_data', [ $this, 'trim_child_order_for_analytics_order_stats' ], 10, 2 );
 
         // remove customer info from order export based on setting
-        add_filter( 'dokan_csv_export_headers', [ $this, 'hide_customer_info_from_vendor_order_export' ], 20, 1 );
+        // add_filter( 'dokan_csv_export_headers', [ $this, 'hide_customer_info_from_vendor_order_export' ], 20, 1 );
```
Comment (mrabbani) [inline includes/Order/MiscHooks.php:33]: We may have to redo this hoook.

**Review by mrabbani — COMMENTED** [review]
Todo
- [ ] Run migration to resync the analytics Data.
- [ ] What will be happen if the dokan stats table is not created but trying to sync the order.

**Comment by shashwatahalder01** [comment]
@mrabbani , bhai kindly resolve the merge conflict

**Comment by shashwatahalder01** [comment]
Scenario:
It is showing 'N/A' instead of default values on analytics. 

Screenshot:
Screenshot at Oct 22 12-36-23
Screenshot at Oct 22 12-10-52

**Comment by shashwatahalder01** [comment]
Scenario:
- [ ] The coupon amount is inconsistent in the Analytics Revenue menu & coupon menu 
- [ ] Multivendor sub-orders are not included in the coupon order count.

Steps to reproduce:
1. place a single vendor order with a coupon
2. multi-vendor order with a coupon 
3. go to analytics > revenue & coupons value [see coupon amount is different]

Screenshot:
Screenshot at Oct 23 16-54-34
Screenshot at Oct 23 16-55-18

**Pull request merged**